spec/controllers/tasks_controller_spec.rb

```diff
@@ -0,0 +1,289 @@
+      it { is_expected.to have_http_status(302) }
+
+      it 'redirects to sign_in' do
+        subject
+
+        expect(response.status).to redirect_to('/users/sign_in')
+      end
+    end
+  end
+
+  describe 'GET#show' do
+    subject { get(:show, id: task_id) }
+
+    context 'when user is authenticated' do
+      let(:user) { create(:user, tasks: [task1, task2]) }
+
+      let(:task1) { create(:task) }
+      let(:task2) { create(:task) }
+      let(:task3) { create(:task) }
+
+      before { sign_in(user) }
+
+      context 'when user has task with provided id' do
+        let(:task_id) { task2.id }
+
+        it { is_expected.to have_http_status(200) }
+        it { is_expected.to render_template(:show) }
+
+        it 'assigns tasks with provided id to instance variable' do
+          subject
+
+          expect(controller.instance_variable_get(:@task)).to eq(task2)
+        end
+      end
+
+      context 'when user does not have task with provided id' do
+        let(:task_id) { task3.id }
+
+        it { is_expected.to have_http_status(400) }
+
+        it 'returns json body' do
+          subject
+
+          expect(JSON.parse(response.body)).to eql(error_response)
+        end
+      end
+    end
+
+    context 'when user is not authenticated' do
+      let(:task_id) { 0 }
+
+      it { is_expected.to have_http_status(302) }
+
+      it 'redirects to sign_in' do
+        subject
+
+        expect(response.status).to redirect_to('/users/sign_in')
+      end
+    end
+  end
+
+  describe 'GET#new' do
+    subject { get(:new, task: task_params) }
+
+    context 'when user is authenticated' do
+      let(:user) { create(:user) }
+      let(:task_params) { FactoryGirl.attributes_for(:task) }
+
+      before { sign_in(user) }
+
+      it 'builds new task', :aggregate_failures do
+        subject
+
+        task = controller.instance_variable_get(:@task)
+        expect(task.user_id).to eq(user.id)
+        expect(task.title).to eq(task_params[:title])
+        expect(task.body).to eq(task_params[:body])
+      end
+
+      it { is_expected.to have_http_status(200) }
+      it { is_expected.to render_template(:new) }
+    end
+
+    context 'when user is not authenticated' do
+      let(:task_params) { {} }
+
+      it { is_expected.to have_http_status(302) }
+
+      it 'redirects to sign_in' do
+        subject
+
+        expect(response.status).to redirect_to('/users/sign_in')
+      end
+    end
+  end
+
+  describe 'POST#create' do
+    subject { post(:create, task: task_params) }
+
+    context 'when user is authenticated' do
+      let(:user) { create(:user) }
+      let(:task_params) { FactoryGirl.attributes_for(:task) }
+
+      before { sign_in(user) }
+
+      it 'creates new task' do
+        expect { subject }.to change { Task.count }.from(0).to(1)
+      end
+
+      it { is_expected.to have_http_status(302) }
+      it { is_expected.to redirect_to(task_path(id: Task.last.id)) }
```

Comment:
let's use `instance_variable_get(:@task)` instead of finding the last task

(this adds an unnecessary dependency on tasks order, for example)